### ISSUE TITLE
Declarations from extensions cannot be overriden yet

### ISSUE THREAD

**Issue by ddmus** [danthorpe/Operations#219]
I've just updated to 2.7.0 and received avalanche of errors from Xcode that I cannot override finished(errors: [ErrorType]) function which is declared newly in Operation class extension. Am I missing something here?

This also applies to updated operationQueue(queue: OperationQueue, didFinishOperation operation: NSOperation, withErrors errors: [ErrorType]) in GroupOperation extension and there might be more. What is now the right approach to override these methods in our project? Thank you.

**Comment by jshier** [comment]
Yes, this is a fundamental limitation of the Swift language right now. I'm guessing it was an oversight in the refactor, in an attempt to clean things up by splitting the large `Operation` class into extensions. As the comments still talk about subclassing, this seems to be a bug.

**Comment by ddmus** [comment]
Ok, I also like to split classes with extensions for readability if it does not break the whole project :-) I would assume such a refactor would go through some testing first and the issue would be spotted earlier.

I wonder how can I overcome this now as we use the framework in CI in combination with Carthage. I guess the best is to roll back to version 2.6 in the cartfile and wait for the update. Do you see any other options? Thanks.

**Comment by danthorpe** [comment]
Hi @ddmus and @jshier - this is totally my mistake, sorry.

I'm gonna fix this by changing `Operation` back to its monolithic class, and removing that rule from SwiftLint. I'll do this now (this morning) and release v2.7.1 later today.

I'll also use this issue as a postmortem to track how this happened and wasn't caught be the tests and figure out how to stop it from happening again.

**Comment by StephenDaedalus** [comment]
For the interim, I just moved the override of `finished(errors: [ErrorType])` into the base class and it seems to have solved all my compilation issues.

**Comment by danthorpe** [comment]
This has been fixed in 2.7.1 now.

**Comment by ddmus** [comment]
Hi @danthorpe, thanks for the quick fix. I've updated to 2.7.1., but I can still see the issue in GroupOperation for didFinishOperation, willFinishOperation and willAddOperation. Can you please check this extension as well? Thank you.

**Comment by danthorpe** [comment]
Yeah, I've just noticed this - totally my fault. I fixed the `Operation` one on the train the other day - didn't realize that I'd also put these into an extension. I'm gonna fix it right now... 2.7.2 impending....

**Comment by danthorpe** [comment]
Although actually... this is not correct. For `GroupOperation` you don't want to be overriding `func operationQueue(queue: OperationQueue, willFinishOperation operation: NSOperation, withErrors errors: [ErrorType])`, instead you should override `func willFinishOperation(operation: NSOperation, withErrors errors: [ErrorType])`.

Essentially the `GroupOperation` is the queue's delegate, and it implements those methods (the ones which receive the queue). In those implementations, we inspect the operation and then call `willFinishOperationOperation()`, after updating the the aggregate errors.

And these `OperationQueueDelegate` methods have always been in an extension.

**Comment by ddmus** [comment]
Yeah.. I can see it now, I am sorry for that. I was confused by the function name change of `didFinishOperation` as Xcode reported it as unavailable, so I replaced it with one of the delegate methods with the similar name.

We should now be using `willFinishOperation` instead for adding operations to the group's queue if we want to do it after some of the former operation ends. Thanks for such a quick response.

**Comment by danthorpe** [comment]
Yeah - that's exactly correct. I did add `@availablity()` annotation for the renamed method - with a replacement. This was to try help people migrate, as Xcode should have offered a "Fix It" suggestion. But, I don't think this always works when the function has more than one argument. Anyway, 2.7.1 should be all good now. 

Gonna leave this issue open for the moment, as I want to look at something things that can be done to stop a regression like this from happening in the future.

**Comment by danthorpe** [comment]
Okay, so this is what happened with this bug.

I added Swift Lint to the project as part of 2.7.0, in both the Xcode build phases and on CI. One of the linting rules is the size of types. And in general I do think that this is a worthwhile rule. However, the `Operation` class is actually pretty big, as it supports dependencies, observers, conditions, cancellation, state management and finishing. I don't or can't really refactor it that much at the moment to break it up, although this is certainly a possibility I'm going to consider in the near future.

Therefore, I organized the class a bit better using `extension`s on itself - and I accidentally moved the `finished(errors: [ErrorType])` function into an extension - which means you cannot subclass it, which is the whole purpose of the function.

So how come this wasn't caught before release? Simply put - I pretty much never use this, and there isn't a unit test which checks that it gets called - there should be. So, I've added an override to this method in `TestOperation` which means that the test target won't compile if it gets moved again. I still haven't yet added a unit test - but I will.

But, also in general, it's pretty tricky to run integration tests for a framework. I have no idea how the framework is actually being consumed. I try to minimize the changes as much as possible, but ultimately I want to push the project forward and keep improving it.

To this end, I have got some example projects which use all my open source frameworks, and I'm going to figure out a way of triggering builds of them in CI when changes occur in one of my upstream repos. The tricky part of this would be how to run Carthage or CocoaPods (or SwiftPM) using a particular commit of a dependencies. In other works, feature branch on Operations framework triggers a build for that commit which also builds & tests the Permissions example app using the same commit - which would require it to be possible to update the dependencies to versions from outside of the repo. I'm thinking something like a `Cartfile.template`

``` 
github "danthorpe/Operations" ${OPERATIONS_HASH}
```

Or maybe to use `--use-submodules` instead - and then configure the CI scripts checkout the specific hash for the submodule. Any other ideas on this are welcome though!
